### DISCUSSION TITLE
How does copilot deals with secrets, passwords, keys etc which we ocassionaly type in our code

### DISCUSSION THREAD

**Discussion by arihantdaga** [community/community#7462] [Copilot Conversations]
Mostly we keep secrets in environment variables. But sometimes we are writing a json key file or editing an environment file or simply writing passwords during testing in one of the file where copilot is active. 
WIll copilot be trained on this text too ? Can it suggest these passwords to some other users ?

I saw copilot auto-completing with a random secret once.  

So i am concerne, if it is safe to use copilot when we tend to accidentally write passwords/secrets in the code during development.

**Comment by martin-marko** [comment]
I'm curious as well, I've seen tokens exposed several times by now, didn't test them but they definitely looked real. What happens if we have a file with a credential in it? Copilot definitely has access to it. This is my only concern with Copilot so far, otherwise, I'm loving everything about it

**Comment by MulleDK19** [comment]
It doesn't learn anything from your code, and your code and data is not used for training.

Any keys and secrets it spits out are either already publicly available, or entirely made up by Copilot.

**Reply by SmallLemur** [reply]
@MulleDK19  this is correct, Copilot is not continuously learning and has been trained on a fix data set coming from public repo. We are looking into features to improve the security of the code generated as well.

**Reply by nvegater** [reply]
Any official reference to this ?

**Reply by SmallLemur** [reply]
Yes "GitHub Copilot is powered by OpenAI Codex, a new AI system created by OpenAI. It has been trained on a selection of English language and source code from publicly available sources, including code in public repositories on GitHub." This is coming from our official FAQ: https://copilot.github.com/
We are working on various features that will improve the security and privacy of copilot. We will announce more soon on our public roadmap

**Reply by nvegater** [reply]
https://docs.github.com/en/github/copilot/github-copilot-telemetry-terms 

What I understand from this link is that it learns from my code as well: 
 _This information may include personal data, including your User Personal Information, as defined in the GitHub Privacy Statement._

**Reply by nvegater** [reply]
It would be nice to have an official reference (like the terms and conditions) to the fact that github copilot wont share my code to the github servers. I understand that it learns from public repositories hosted in the Github servers. But its unclear wether it continues learning from the places where its used (e.g. my Jetbrains IDE using copilot plugin).

**Reply by SmallLemur** [reply]
I will look into it and get back to you. It might take a few days with Thanksgiving at the moment.

**Reply by nvegater** [reply]
Thank you very much! Looking forward to your answer.

**Reply by SmallLemur** [reply]
Apologies for not answering before. The telemetry data are captured only when Copilot is on. During the technical Preview, it is the only way of not sharing telemetry data with Copilot. We are working on features in future that would allow for more control.

**Comment by ib0b** [comment]
Code snippets, accepted suggestions and edited suggestions are sent to the Copilot team. Read more about their telemetry policy here: https://docs.github.com/en/github/copilot/about-github-copilot-telemetry

So you should not use copilot on such sensitive projects, but from a training perspective, it won't have access to your code. Though copilot engineers might.
